--- node-v0.4.11/wscript	2011-08-18 10:18:10.000000000 +0900
+++ files/wscript	2011-10-04 09:33:49.000000000 +0900
@@ -440,10 +440,9 @@
     if 'DEST_CPU' in conf.env:
       arch = conf.env['DEST_CPU']
       # map supported_archs to GCC names:
-      arch_mappings = {'ia32': 'i386', 'x64': 'x86_64'}
+      arch_mappings = {'ia32': '-m32', 'x64': '-m64'}
       if arch in arch_mappings:
-        arch = arch_mappings[arch]
-      flags = ['-arch', arch]
+        flags = arch_mappings[arch]
       conf.env.append_value('CCFLAGS', flags)
       conf.env.append_value('CXXFLAGS', flags)
       conf.env.append_value('LINKFLAGS', flags)
